src/gp_doctest.cpp

```diff
@@ -2760,6 +2760,188 @@ TEST_CASE("TPEngine: Branch Length Optimization") {
+      for (auto& tree : trees) {
+        bool tree_found = false;
+        for (size_t i = 0; i < tree_map_2.size(); i++) {
+          if (tree == tree_map_2[i]) {
+            tree_found = true;
+            break;
+          }
+        }
+        if (!tree_found) {
+          tree_map_2.push_back(tree);
+        }
+      }
+    }
+
+    std::string newick_1;
+    for (auto& tree : tree_map_1) {
+      newick_1 += tree.Topology()->Newick() + "\n";
+    }
+    std::string newick_2;
+    for (auto& tree : tree_map_2) {
+      newick_2 += tree.Topology()->Newick() + "\n";
+    }
+    return (newick_1 == newick_2);
+  };
+
+  // Export a spanning newick file from TPEngine, then build new DAG from that file.
+  // Compare to the input DAG.
+  auto BuildCoveringNewickAndCompareNewDAG = [](GPInstance& inst_1) {
+    const std::string temp_newick_path = "_ignore/temp.newick";
+    std::ofstream file_out;
+    file_out.open(temp_newick_path);
+    file_out << inst_1.GetDAG().ToNewickOfCoveringTopologies() << std::endl;
+    file_out.close();
+
+    auto inst_2 = GPInstanceOfFiles(inst_1.GetFastaSourcePath(), temp_newick_path,
+                                    "_ignore/mmapped_pv.data1");
+    return SubsplitDAG::Compare(inst_1.GetDAG(), inst_2.GetDAG(), false);
+  };
+
+  // Export a top tree newick file from TPEngine, then build new TPEngine from that
+  // file. Compare to the input TPEngine.
+  auto BuildTopTreeNewickAndCompareNewTPEngine =
+      [&BuildTopTreeMapViaBruteForce](GPInstance& inst_1) {
+        std::ofstream file_out;
+        const std::string temp_newick_path_1 = "_ignore/temp_1.newick";
+        file_out.open(temp_newick_path_1);
+        file_out << inst_1.GetTPEngine().ToNewickOfTopTrees() << std::endl;
+        file_out.close();
+        const std::string temp_newick_path_2 = "_ignore/temp_1.newick";
+        file_out.open(temp_newick_path_2);
+        const auto tree_map = BuildTopTreeMapViaBruteForce(inst_1);
+        for (const auto& tree : tree_map) {
+          file_out << inst_1.GetDAG().TreeToNewickTopology(tree) << std::endl;
+        }
+        file_out.close();
+
+        auto inst_2 = GPInstanceOfFiles(inst_1.GetFastaSourcePath(), temp_newick_path_2,
+                                        "_ignore/mmapped_pv.data1");
+        inst_2.MakeTPEngine();
+        inst_2.MakeNNIEngine();
+        return TPEngine::Compare(inst_1.GetTPEngine(), inst_2.GetTPEngine(), false);
+      };
+
+  auto inst_1 =
+      GPInstanceOfFiles(fasta_path, newick_path_1, "_ignore/mmapped_pv.data1");
+  auto inst_2 =
+      GPInstanceOfFiles(fasta_path, newick_path_2, "_ignore/mmapped_pv.data2");
+  inst_1.MakeTPEngine();
+  inst_1.MakeNNIEngine();
+  auto& tp_engine = inst_1.GetTPEngine();
+  auto& nni_engine = inst_1.GetNNIEngine();
+  inst_2.MakeTPEngine();
+  inst_2.MakeNNIEngine();
+
+  CHECK_MESSAGE(
+      TPEngine::Compare(inst_1.GetTPEngine(), inst_1.GetTPEngine(), false) == 0,
```

Comment:
Do you mean to have inst_1 and inst_2 here?

Reply:
Or are you actually testing the Compare functionality?

Reply:
Yes it is testing compare functionality.